### PR TITLE
V3 monitoring jobs - add new queue based ops tools

### PR DESCRIPTION
**Pull request by scottbommarito**
Depends on https://github.com/NuGet/NuGet.Services.Metadata/pull/244

Added `Feed2Monitoring` and `Catalog2MonitoringReprocessing`, two ops tools that can alternatively be set up to run regularly.

`Feed2Monitoring`:
Reads all packages from the feed and verifies that they have been checked by the V3 monitoring jobs. Can be run if there are concerns that the V3 monitoring jobs have skipped packages.

`Catalog2MonitoringReprocessing`:
Reads the entire catalog (using lightning) and queues the latest commit for each package. Should be run to regenerate the monitoring status blobs (instead of running `Catalog2Monitoring`).

### CODE REVIEW

File: src/Ng/Jobs/Catalog2MonitoringReprocessingJob.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+using NuGet.Packaging.Core;
+using NuGet.Services.Configuration;
+using NuGet.Services.Metadata.Catalog;
+using NuGet.Services.Metadata.Catalog.Helpers;
+using NuGet.Services.Metadata.Catalog.Monitoring;
+using NuGet.Services.Storage;
+
+namespace Ng.Jobs
+{
+    public class Catalog2MonitoringReprocessingJob : LoopingNgJob
```
Comment (loic-sharma) [inline src/Ng/Jobs/Catalog2MonitoringReprocessingJob.cs]: >Catalog2MonitoringReprocessingJob [](start = 17, length = 33)

Please add a description to what these jobs do here.

File: src/Ng/Jobs/Catalog2MonitoringReprocessingJob.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+            Logger.LogInformation("Processing packages.");
+            foreach (var packageEntry in packageEntries)
+            {
+                var feedPackage = new FeedPackageIdentity(new PackageIdentity(packageEntry.Id, packageEntry.Version));
+                Logger.LogInformation("Processing {Id} {Version}", feedPackage.Id, feedPackage.Version);
+
+                if (!validatedPackages.Contains(feedPackage))
+                {
+                    Logger.LogError("{Id} {Version} was skipped!", feedPackage.Id, feedPackage.Version);
```
Comment (loic-sharma) [inline src/Ng/Jobs/Catalog2MonitoringReprocessingJob.cs]: >LogError [](start = 27, length = 8)

This isn't an error, right? Considering changing this to `LogWarning`. Also, maybe add something like "Reprocessing package" to the log message.

File: src/Ng/Jobs/Catalog2MonitoringReprocessingJob.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                        new PackageValidatorContext(feedPackage, new CatalogIndexEntry[] { packageEntry }),
+                        cancellationToken);
+                }
+            }
+        }
+
+        private async Task<HashSet<FeedPackageIdentity>> GetValidatedPackages(CancellationToken cancellationToken)
+        {
+            var validatedPackagesList = await _statusService.ListAsync(cancellationToken);
+            return new HashSet<FeedPackageIdentity>(validatedPackagesList.Select(p => p.Package).ToList());
```
Comment (loic-sharma) [inline src/Ng/Jobs/Catalog2MonitoringReprocessingJob.cs]: >ToList [](start = 97, length = 6)

Is the `ToList` necessary? This forces the `validatedPackagesList` to be iterated twice: once for the `ToList`, and a second time to add each package to the `HashSet`.

File: src/Ng/Jobs/Feed2MonitoringJob.cs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+            var invalidPackages = new ConcurrentQueue<FeedPackageIdentity>();
+            _pagesProcessed = 0;
+
+            await ParallelExecutionUtility.ProcessInParallel(() => ProcessPackages(validPackagesSet, invalidPackages, cancellationToken), _degreeOfParallelism);
+
+            if (!invalidPackages.IsEmpty)
+            {
+                var invalidPackagesList = new List<FeedPackageIdentity>();
+                Logger.LogError("The following packages are missing:");
```
Comment (loic-sharma) [inline src/Ng/Jobs/Feed2MonitoringJob.cs]: >LogError [](start = 23, length = 8)

I think that seeing a bunch of error logs with just ids and versions may be confusing. Maybe change the first `LogError` to "<X> packages are missing." and the second `LogError` message to "{id} {version} is missing."

File: src/Ng/Jobs/Feed2MonitoringJob.cs [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                while (invalidPackages.TryDequeue(out var invalidPackage))
+                {
+                    invalidPackagesList.Add(invalidPackage);
+                    Logger.LogError($"{invalidPackage.Id} {invalidPackage.Version}");
```
Comment (loic-sharma) [inline src/Ng/Jobs/Feed2MonitoringJob.cs]: >LogError [](start = 27, length = 8)

I wouldn't format the log strings - loggers often have special optimizations/laziness for string formatting.

File: src/Ng/Jobs/Feed2MonitoringJob.cs [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        {
+            var requestSuccessful = false;
+            var failures = 0;
+
+            T result = null;
+            while (!requestSuccessful || failures > _maxRequestFailures)
+            {
+                try
+                {
+                    result = await action();
```
Comment (loic-sharma) [inline src/Ng/Jobs/Feed2MonitoringJob.cs]: >result [](start = 20, length = 6)

Nit: you could just return the result of the action directly, get rid of the `requestSuccessful` variable, and return `null` on line 187.

**Comment by scottbommarito** [comment]
Decided to close this PR because it's not currently necessary and would require work to fix. In the future when these ops tools are needed, we can look back into this branch and fix it.

**Pull request closed without merging**